FR #962: Add Correction Handling and Delivery Handling attributes and refactor Invoice Handling (#1501)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/glossary.md

The time window for which a charge is effective, inclusive of the start date and exclusive of the end date. The charge period for continuous usage should match the time granularity of the dataset (e.g., 1 hour for hourly, 1 day for daily). The charge period for a non-usage charge with time boundaries should match the period of eligibility.

<a name="glossary:closed-billing-period"><b>Closed Billing Period</b></a>

> A [*billing period*](#glossary:billing-period) with [Billing Period Status](#datasets.billingperiod.billingperiodstatus) set to "Closed". The period has been financially closed after all planned invoices for the period have been [*issued*](#glossary:issued-invoice) by designated [*invoice issuers*](#glossary:invoice-issuer) and no additional invoices will be associated with this period. Exceptionally, additional invoices may be associated with a closed billing period if explicitly requested by the customer.

Reviewer: This seems too restrictive. A customer requesting an invoice for a particular period seems like an odd exception. Do we know of any providers that let the customer change which billing period something gets charged in (after the fact)? 

Isn't the more common scenario here that the invoice issuer discovered some correction (outage credits, etc.) and wants to true up the incident in the billing period without waiting for another period? If payment terms are net 30/45 etc., some provider may not want to wait to apply the change till the current billing period ending to true-up. Or is the thinking that invoice issuers cannot provide an invoice after close unless the customer manually requests this? Seems not practical.

Reviewer 2: good catch!

I missed aligning this with the current version of the requirement 😞
I’ve just updated it - replaced `...may be associated with a closed billing period if explicitly requested by the customer` with `...may be associated with a closed billing period if explicitly requested or approved by the customer.`

@udam-stitcher Does that work for you?

For context, the goal here is to “protect” closed billing periods - once a BillingPeriodStatus is set to Closed, new charge records should not be added and new invoices should not be issued for that billing period, unless explicitly requested or approved by the customer.

Reviewer: Sounds good in theory but I don't know how this would work. What's this approval process you're referring to? Im not aware of any process like this. What i was getting at was the changes are typically due to things the Invoice Issuer wants to get out, not what a customer approves. So yes, if a provider is happy to hold the credit or extra charges for another billing period, they they might. Most likely though they will want to pass those charges to the customer so they can get paid.

Author: @udam-stitcher This language came out of conversations in TF-1 with @pydi-aws and others, to be inclusive enough to cover blanket understandings / agreements communicated from customer to provider to always follow this exception.

The important point is that the customer is somehow involved in approving that the process operates differently than what we have written here.  In which case, fine, all bets are off, and closed billing periods will operate differently than documented here.

Reviewer 2: The transition protection requirement ("MUST NOT transition from 'Closed' to 'Open' unless explicitly requested or approved by the customer") may not be honored by large providers when dealing with smaller customers. 

However, smaller providers already effectively respect this today — once they declare all invoices have been issued for a period, they don't unilaterally issue additional ones. 😄 

Attaching relevant content either already merged or proposed in this PR:

- **(*updated in this PR*)** BillingPeriodStatus Allowed Value "Closed": "The billing period has ended, all anticipated invoices have been issued, and the delivered data is finalized."
- **(*merged*)** InvoiceDetailCreated for a given BillingPeriodStart and InvoiceIssuerName MUST be earlier than or equal to BillingPeriod.BillingPeriodLastUpdated for the same BillingPeriod.BillingPeriodStart and BillingPeriod.InvoiceIssuerName when BillingPeriod.BillingPeriodStatus is "Closed".
- **(*proposed in this PR*)** BillingPeriodStatus MUST NOT transition from "Closed" to "Open" unless explicitly requested or approved by the customer.

Author: Discussed in Mar 24 TF-1 and agreed that the language in this glossary entry merely reflects the relevant normative requirements proposed in this PR.

@udam-stitcher Please leave a suggestion if you wish to tweak the verbiage.  Otherwise, no action here.

Reviewer: If you guys are aligned on this, don't want to stand in the way. I just think we're getting well outside the data spec realm with this, but if there's alignment in this direction, let's go for it.

Author: Discussed at length in Mar 31 TF-1.  The group agreed that the language here aligns with expectations communicated in the normative requirements as previously agreed in TF-1 with @pydi-aws and others.  

Resolving with no action.  Feel free to open and provide a specific suggestion if you wish.

> A [*billing period*](#glossary:billing-period) with [Billing Period Status](#datasets.billingperiod.billingperiodstatus) set to "Closed". The period has been financially closed after all planned invoices for the period have been [*issued*](#glossary:issued-invoice) by designated [*invoice issuers*](#glossary:invoice-issuer) and no additional invoices will be associated with this period. Exceptionally, additional invoices may be associated with a closed billing period if explicitly requested by the customer.

Reviewer 2: Just noticed that I missed aligning this with the current version of the requirement:

```suggestion
A [*billing period*](#glossary:billing-period) with [Billing Period Status](#datasets.billingperiod.billingperiodstatus) set to "Closed". The period has been financially closed after all planned invoices for the period have been [*issued*](#glossary:issued-invoice) by designated [*invoice issuers*](#glossary:invoice-issuer) and no additional invoices will be associated with this period. Exceptionally, additional invoices may be associated with a closed billing period if explicitly requested or approved by the customer.
```

<a name="glossary:correction"><b>Correction</b></a>

Any modification (including updates, additions, or omissions) to previously delivered records within a defined [*delivery scope*](#glossary:delivery-scope) (e.g., temporal grouping such as a [*billing period*](#glossary:billing-period) or non-temporal, logical grouping such as a [*contract*](#glossary:contract)).

> A correction (lowercase) may consist of one or more simultaneous changes, including updates to or omission of previously delivered records, or the addition of new records that supplement previously delivered data within the affected *delivery scope*. This concept applies across all FOCUS datasets.

Author: ### Missing Attribute Bindings for Other Datasets
The diff introduces two new attributes: `CorrectionHandling` and `DeliveryHandling`.
* In `specification/datasets/cost_and_usage/dataset.md` and `specification/datasets/contract_commitment/dataset.md`, you explicitly added the requirements that they `MUST conform to [CorrectionHandling]` and `[DeliveryHandling]`.
* **The Issue:** You *didn't* add these requirements to the `Invoice Detail` or `Billing Period` datasets. However, the new glossary entry for `Correction` explicitly states: *"This concept applies across all FOCUS datasets,"* and both datasets are heavily involved in the delivery and correction concepts outlined in the new appendix. You either need to update `specification/datasets/invoice_detail/dataset.md` and `specification/datasets/billing_period/dataset.md` to require conformance to these new attributes, or you need to revise the glossary entry to not say that Corrections are applicable across all FOCUS datasets.

Reviewer 2: Thanks! I believe this is addressed in the updated version.

---

### Overview:

Cost and Usage (specification/datasets/cost_and_usage/dataset.md)
- Removed InvoiceHandling conformance requirement
- Added CorrectionHandling and DeliveryHandling conformance requirements
- Reordered conformance requirements

Invoice Detail (specification/datasets/invoice_detail/dataset.md)
- Removed InvoiceHandling conformance requirement
- Added CorrectionHandling and DeliveryHandling conformance requirements

Billing Period (specification/datasets/billing_period/dataset.md)
- Added CorrectionHandling and DeliveryHandling conformance requirements

Contract Commitment (specification/datasets/contract_commitment/dataset.md)
- Added CorrectionHandling and DeliveryHandling conformance requirements

Author: Looks good!  Resolving.

A physical representation of a specific [*dataset instance*](#glossary:dataset-instance) delivered by a [data generator](#metadata.datagenerator).

<a name="glossary:delivery-scope"><b>Delivery Scope</b></a>

> A dataset-specific boundary or set of boundaries that determines which records are included in a [*dataset artifact*](#glossary:dataset-instance-artifact) delivery. Scopes can be temporal (e.g., a [*billing period*](#glossary:billing-period) or non-temporal (e.g., a [*contract*]((#glossary:contract)) or other logical grouping), and multiple scopes may be applicable for a single dataset depending on use case or delivery configuration. Scopes determine how Overwrite and Append *dataset artifact* deliveries, as well as corrections, are applied.

Author: ```suggestion
A dataset-specific boundary or set of boundaries that determines which records are included in a [*dataset artifact*](#glossary:dataset-instance-artifact) delivery. Scopes can be temporal (e.g., a [*billing period*](#glossary:billing-period) or non-temporal (e.g., a [*contract*](#glossary:contract) or other logical grouping), and multiple scopes may be applicable for a single dataset depending on use case or delivery configuration. Scopes determine how Overwrite and Append *dataset artifact* deliveries, as well as corrections, are applied.
```

Reviewer 2: good catch! thanks!

<a name="glossary:FOCUS-dataset"><b>FOCUS Dataset</b></a>

A structured collection of columns that conforms to the BCP14 criteria established by FOCUS. All columns included must be defined in the FOCUS Columns section of the specification.

In addition to these standardized columns, data generators may include custom provider-specific columns (prefixed with `x_`) where additional context is needed beyond what is captured in the defined FOCUS columns. If custom provider-specific columns introduce record-splitting (i.e., a single original charge results in multiple rows), the data generator is responsible for ensuring that all cost and quantity metrics still meet the aggregation and consistency rules required by the specification.

The collection of datasets are designed to provide billing insight, additional context, metadata, mapping, or enrichment information that enhances the interpretability or completeness.

<a name="glossary:inclusivestartbound"><b>Inclusive Start Bound</b></a>

A Date/Time Format value that is contained within the beginning bound of a time period.

<a name="glossary:invoice"><b>Invoice</b></a>

> A formal document that reflects the charges for resources or services consumed by a customer.

Author: ```suggestion
A document that summarizes the charges for resources or services consumed by a customer.
```

<a name="glossary:invoice-reconciliation"><b>Invoice Reconciliation</b></a>

> The process of ensuring that the aggregated cost and usage information presented on an [*invoice*](#glossary:invoice) matches the detailed cost and usage [*charges*](#glossary:charge) presented in a [*FOCUS dataset](#glossary:FOCUS-dataset).

Author: ```suggestion
The process of ensuring that the aggregated cost and usage information presented on an [*invoice*](#glossary:invoice) matches the detailed cost and usage [*charges*](#glossary:charge) presented in a [*FOCUS dataset*](#glossary:FOCUS-dataset).
```

A Date/Time Format value that is contained within the beginning bound of a time period.

<a name="glossary:invoice"><b>Invoice</b></a>

A document that summarizes the charges for resources or services consumed by a customer.

<a name="glossary:invoice-reconciliation"><b>Invoice Reconciliation</b></a>

> The process of ensuring that aggregated Cost and Usage [*FOCUS dataset*](#glossary:FOCUS-dataset) records match the corresponding Invoice Detail [*FOCUS dataset*](#glossary:FOCUS-dataset) records, and that both align with the associated cost and usage information presented on an [*invoice*](#glossary:invoice).

Reviewer: Not really about Invoice Detail.. this is a general process in the industry of reconciling the detailed data (cost or sometimes just based on usage). I think we might say it's supported by Invoice Detail, but it's not really what the general invoice reconciliation process is about.

Reviewer 2: I hear you. Would this updated version work for you?

First paragraph defines invoice reconciliation as a general industry process, second paragraph describes how it applies in the context of FOCUS.

> <a name="glossary:invoice-reconciliation"><b>Invoice Reconciliation</b></a>
> 
> The process of ensuring that detailed cost and usage data is accurate, complete, and matches across all relevant sources, typically involving the reconciliation of raw usage or cost records against invoiced amounts.
> 
> In the context of FOCUS, this process involves ensuring consistency between the [*invoice*](#glossary:invoice), the Invoice Detail [*FOCUS dataset*](#glossary:FOCUS-dataset), and the Cost and Usage *FOCUS dataset*, specifically, that aggregated data across all three sources is aligned.

Reviewer: i would say its more the opposite - its the process of ensuring the invoices are correct by reconciling with the detailed cost and/or usage information available to you. It's also a process done by the end customer not a validation done by the data generators necessarily.

Author: @udam-stitcher 
Hmm... I see @ijurica's suggestion as fulfilling your concern.  Could you provide a suggestion that you'd prefer?

Reviewer: It still sounds to me like 'detailed cost and usage data reconciliation', not 'invoice reconciliation', but it's really a minor comment, and I'm ok if you guys are aligned on this.

Heres a suggestion:

Invoice Reconciliation

The process of verifying that billed amounts and metering units in a vendor invoice match actual or contracted costs across internal systems, external data sources, and agreements. In the context of FOCUS, this process reconciles cost and usage information across the vendor invoice, the Invoice Detail dataset, and the Cost and Usage dataset.

Author: Discussed at length in Mar 31 TF-1.  While the group preferred the existing verbiage, there were elements of the story Udam is trying to tell in his version that I don't want to lose.  So how about:

> The process of verifying that the costs, quantities, and applied adjustments on an invoice match the underlying usage records and contracted rates for a specific billing period.
> 
> In the context of FOCUS, this process ensures consistency by reconciling cost and usage data across the invoice, the Invoice Detail dataset, and the Cost and Usage dataset to identify and resolve any discrepancies.

@ijurica Does that work for you?  I have suggested this via [this thread](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1501/files#r3016859099).

Author: Accepted suggestion via that thread mentioned in my previous comment.  Resolving this thread.

A Date/Time Format value that is contained within the beginning bound of a time period.

> <a name="glossary:invoice"><b>Invoice</b></a>

Author: ### Glossary Alphabetical Sorting Error
The new glossary terms were inserted in a way that breaks alphabetical order:
* **The Issue:** `Invoice` and `Invoice Reconciliation` were placed *before* `Interruptible`. 
* **The Fix:** Alphabetically, `I-n-t` (Interruptible) comes before `I-n-v` (Invoice). `Interruptible` needs to be moved above the new `Invoice` entries to maintain correct sorting in `specification/glossary.md`.

Reviewer 2: Addressed positioning of he glossary terms introduced in this PR. 

I do notice that some existing terms are not fully aligned, but I’ve intentionally avoided adjusting those here to minimize conflicts. We typically address this during the Consistency Review phase.

Author: Sounds good, resolving.

A document that summarizes the charges for resources or services consumed by a customer.

<a name="glossary:invoice-reconciliation"><b>Invoice Reconciliation</b></a>

> The process of ensuring that aggregated Cost and Usage [*FOCUS dataset*](#glossary:FOCUS-dataset) records match the corresponding Invoice Detail [*FOCUS dataset*](#glossary:FOCUS-dataset) records, and that both align with the associated cost and usage information presented on an [*invoice*](#glossary:invoice).

Reviewer 2: ```suggestion
The process of ensuring that detailed cost and usage data is accurate, complete, and matches across all relevant sources, typically involving the reconciliation of raw usage or cost records against invoiced amounts.

In the context of FOCUS, this process is performed by comparing aggregated Cost and Usage [*FOCUS dataset*] records with corresponding Invoice Detail [*FOCUS dataset*] records, and ensuring that both align with the associated cost and usage information presented on an [*invoice*](#glossary:invoice).
```

A category of compute resources that can be paused or terminated by the CSP within certain criteria, often advertised at reduced unit pricing when compared to the equivalent non-interruptible resource.

<a name="glossary:invoice"><b>Invoice</b></a>

A document that summarizes the charges for resources or services consumed by a customer.

<a name="glossary:invoice-reconciliation"><b>Invoice Reconciliation</b></a>

The process of ensuring that detailed cost and usage data is accurate, complete, and matches across all relevant sources, typically involving the reconciliation of raw usage or cost records against invoiced amounts.

> In the context of FOCUS, this process involves ensuring consistency between the [*invoice*](#glossary:invoice), the Invoice Detail [*FOCUS dataset*](#glossary:FOCUS-dataset), and the Cost and Usage *FOCUS dataset*, specifically, that aggregated data across all three sources is aligned.

Author: Per discussion of [this thread](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1501/files/a29570ab1ab9240171fd06c3e192766cf1614a0c#r3012802070):

```suggestion
The process of verifying that the costs, quantities, and applied adjustments on an [*invoice*](#glossary:invoice) match the underlying usage records and contracted rates for a specific [*billing period*](#glossary:billing-period).

In the context of FOCUS, this process ensures consistency by reconciling cost and usage data across the invoice, the [Invoice Detail](#datasets.invoicedetail) dataset, and the [Cost and Usage](#datasets.costandusage) dataset to identify and resolve any discrepancies.
```

<a name="glossary:issued-invoice"><b>Issued Invoice</b></a>

> An [*invoice*](#glossary:invoice) that has been formally [*reconciled*](#glossary:invoice-reconciliation) and issued by the designated invoice issuer ([Invoice Issue Status](#datasets.invoicedetail.invoiceissuestatus) set to "Issued"). Once issued, the invoice becomes the authoritative financial document and is considered finalized and the financial data presented on the invoice must not be altered.

Reviewer 3: ```suggestion
An [*invoice*](#glossary:invoice) that has been formally [*reconciled*](#glossary:invoice-reconciliation) and issued by the designated invoice issuer ([Invoice Issue Status](#datasets.invoicedetail.invoiceissuestatus) set to "Issued"). Once issued, the invoice becomes the authoritative financial document and is considered finalized. The financial data presented on the invoice is not expected to change.
```

Two changes: (1) Splits into two sentences for readability. (2) Replaces lowercase "must not be altered" with "is not expected to change" to avoid normative-adjacent language in a glossary definition. The actual enforcement lives in the InvoiceIssueStatus requirements.
